better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+above, are exposed to users of the type, and thus should be in public
+doc comments.  Others, like the invariant that the members of
+`PairArray` have the same length, are purely part of its
+implementation, and should be encoded in ordinary comments addressed
+privately to the implementor/maintainer of the code.  Note that you
+can have both: `PairArray` *also* has a public invariant that its
+`count` is non-negative.  We'll get to why this particular invariant
+is not explicitlty documented in a moment…
+
+### Making It Tractable
+
+If we're going to integrate documentation into our programming so
+tightly, we need to make sure it's neither intrusive for the
+maintainer to read nor burdensome for the author to write.  Ideally,
+it should help both of them.
+
+Now, not every contract is as simple as the ones we've shown so far,
+but simplicity is a goal.  In fact, if you can't write a terse,
+simple, but _complete_ contract for a component, there's a good chance
+it's badly designed.  A classic example is the C library `realloc`
+function, which does at least three different things—allocate, deallocate, and resize 
+dynamic memory—all of which
+need to be described. A better design would have separated these
+functions.  So simple contracts are both easy to digest and easy to
+write.
+
+Let's look at another example. The methodology we'll be following here is based on the [Swift API
+Design
+Guidelines](https://www.swift.org/documentation/api-design-guidelines/),
+only slightly modified.
+
+```swift
+/// A resizable random-access collection of `T`s.
+struct DynamicArray<T> {
+
+  /// Removes and returns the last element.
+  public mutating func popLast() -> T { ... }
+
+  /// The `i`th element.
+  public subscript(i: Int) -> T { ... }
+
+  /// The length.
+  public var length: Int { ... }
+    .
+    .
+    .
+
+  /// The number of elements that can be stored before storage is
+  /// reallocated.
+  private var capacity: Int { ... }
+}
+```
+
+The method bodies are collapsed, because contracts are not about what
+you put in your implementations.  Contracts are part of your
+function's interface.
+
+#### Basics
+
+We've started by filling in the most basic part of every documentation
+comment: a summary sentence fragment that minimally describes what the
+thing *is* or *does*.
+
+The first one
+
+```swift
+/// A resizable random-access collection of `T`s.
+struct DynamicArray<T>
+```
+
+gives us the context we need to understand the methods: we're looking
+at the declaration of dynamic array type that holds any number of Ts.
+
+Now let's look at the documentation for the first method, called
+"popLast."
+
+```swift
+  /// Removes and returns the last element.
+  public mutating func popLast() -> T { ... }
+```
+
+As you can see from the summary, it removes and returns the last
+element. Notice that the phrase “last element” is meaningful only
+because we documented that `DynamicArray` is a collection, which we
+know is a sequence of elements.
```

Comment:
Is it useful to say why we know it's a sequence of elements?

Reply:
In a chapter so focused on _why_ we know the things we know, a phrase like "which we know" feels like it demands some justification. I wouldn't just rely on conventional usage of words. 

Does the Swift API guidelines say anything about how to use english words with widely-accepted definitions?

Reply:
@nickpdemarco [here](https://www.swift.org/documentation/api-design-guidelines/#use-terminology-well).  But you really should know that because you should have read them.  Go read them today please 😃 

@sean-parent I understand your complaint about “`Collection`.”  Maybe we should have called it “`MultipassSequence`”. With regard to your suggested documentation comment, does resizeable imply multipass?  I'm not sure. And regardless of what the best answer is for that name, `Collection` has an established meaning in Swift and it would be weird to bypass it.  So I think we should use `Collection` here.

W.R.T. Nick's original question,  we know it because `Collection` refines `Sequence` but I'm not still not sure it's worth saying that. 🤷